src/transformStateWithClones.js

```diff
@@ -7,7 +7,45 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const copyState = {
+    ...state,
+  };
+
+  const history = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        addProperties(copyState, action.extraData);
+        break;
+      case 'removeProperties':
+        removeProperties(copyState, action.keysToRemove);
+        break;
+      case 'clear':
+        clear(copyState);
+        break;
+    }
+
+    history.push({
+      ...copyState,
+    });
+  }
+
+  return history;
+}
+
+function addProperties(copyState, extraData) {
+  Object.assign(copyState, extraData);
+};
+
+function removeProperties(copyState, keysToRemove) {
+  for (const key of keysToRemove) {
+    delete copyState[key];
+  }
+}
+
+function clear(copyState) {
+  Object.keys(copyState).forEach(key => delete copyState[key]);
```

Comment:
If I remember there is a permission to use object's methods like map, forEach etc. But I don't know, it looks legit for me.

Reply:
> If I remember there is a permission to use object's methods like map, forEach etc. But I don't know, it looks legit for me.

It's ok